File: shared/api_config.ts
import OpenAI from "openai";
import fs from 'fs/promises';
import path from 'path';

export interface Message {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

interface ConversationEntry {
  module: string;
  timestamp: string;
  messages: Message[];
  response: string;
}

interface MasterLog {
  project_start_time: string;
  conversations: {
    [key: string]: ConversationEntry[];
  };
}

export class AIClient {
  private client: OpenAI;
  private conversationHistory: { [key: string]: Message[] };
  private conversationLogDir: string;
  private masterLogPath: string;
  private moduleName: string;
  private masterLog!: MasterLog;

  static async create(moduleName: string): Promise<AIClient> {
    const client = new AIClient(moduleName);
    await client.initialize();
    return client;
  }

  private constructor(moduleName: string) {
    const apiKey = process.env.OPENAI_API_KEY;
    if (!apiKey) {
      throw new Error('OPENAI_API_KEY environment variable is required');
    }

    this.client = new OpenAI({
      baseURL: "https://openrouter.ai/api/v1",
      apiKey: apiKey,
      defaultHeaders: {
        "HTTP-Referer": "http://localhost:3000",
        "X-Title": "Quiz-Show-Game"
      }
    });

    this.moduleName = moduleName;
    this.conversationHistory = {};
    this.conversationLogDir = 'logs/conversations';
    this.masterLogPath = 'logs/master_conversation_log.json';
  }

  private async initialize(): Promise<void> {
    this.masterLog = await this._loadOrCreateMasterLog();
  }

  private async _loadOrCreateMasterLog(): Promise<MasterLog> {
    try {
      await fs.mkdir(path.dirname(this.masterLogPath), { recursive: true });
      try {
        const data = await fs.readFile(this.masterLogPath, 'utf-8');
        return JSON.parse(data);
      } catch (error) {
        // File doesn't exist or is invalid, create new log
        const newLog: MasterLog = {
          project_start_time: new Date().toISOString(),
          conversations: {}
        };
        await fs.writeFile(this.masterLogPath, JSON.stringify(newLog, null, 2));
        return newLog;
      }
    } catch (error) {
      console.error('Error loading/creating master log:', error);
      throw error;
    }
  }

  private async _logConversation(conversationId: string, messages: Message[], response: string) {
    try {
      const timestamp = new Date().toISOString();
      const conversationEntry: ConversationEntry = {
        module: this.moduleName,
        timestamp,
        messages,
        response
      };

      // Ensure conversation log directory exists
      await fs.mkdir(this.conversationLogDir, { recursive: true });

      // Save individual conversation log
      const logPath = path.join(
        this.conversationLogDir,
        `conversation_${conversationId}_${timestamp.replace(/[:.]/g, '-')}.json`
      );
      await fs.writeFile(logPath, JSON.stringify(conversationEntry, null, 2));

      // Update master log
      if (!this.masterLog.conversations[conversationId]) {
        this.masterLog.conversations[conversationId] = [];
      }
      this.masterLog.conversations[conversationId].push(conversationEntry);
      await fs.writeFile(this.masterLogPath, JSON.stringify(this.masterLog, null, 2));
    } catch (error) {
      console.error('Error logging conversation:', error);
      throw error;
    }
  }

  async get_completion(messages: Message[], conversationId?: string, temperature: number = 0.7): Promise<string> {
    if (!Array.isArray(messages) || messages.length === 0) {
      throw new Error('Messages array must not be empty');
    }
    try {
      const conversationMessages = conversationId && this.conversationHistory[conversationId] 
        ? [...this.conversationHistory[conversationId], ...messages]
        : messages;

      const completion = await this.client.chat.completions.create({
        model: "anthropic/claude-3.5-sonnet:beta",
        messages: conversationMessages,
        temperature,
        max_tokens: 1000
      });

      const content = completion.choices[0].message.content || '';

      if (conversationId) {
        if (!this.conversationHistory[conversationId]) {
          this.conversationHistory[conversationId] = [];
        }
        this.conversationHistory[conversationId].push(
          messages[messages.length - 1],
          { role: "assistant" as const, content }
        );
      }

      this._logConversation(
        conversationId || "single_interaction",
        messages,
        content
      );

      return content;

    } catch (error) {
      console.error('Error getting completion:', error);
      throw error;
    }
  }

  startConversation(): string {
    const conversationId = new Date().toISOString();
    this.conversationHistory[conversationId] = [];
    return conversationId;
  }

  endConversation(conversationId: string): void {
    if (conversationId in this.conversationHistory) {
      this._logConversation(
        conversationId,
        this.conversationHistory[conversationId],
        "CONVERSATION_ENDED"
      );
      delete this.conversationHistory[conversationId];
    }
  }

  getConversationHistory(conversationId: string): Message[] {
    return this.conversationHistory[conversationId] || [];
  }

  clearConversationHistory(conversationId?: string): void {
    if (conversationId) {
      if (conversationId in this.conversationHistory) {
        this.conversationHistory[conversationId] = [];
      }
    } else {
      this.conversationHistory = {};
    }
  }
}

File: api_config.py
from openai import OpenAI
import os
from datetime import datetime
import logging
import json

# Configure logging
logging.basicConfig(level=logging.INFO)

def setup_logging(module_name):
    """Setup logging for each module with timestamps"""
    timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
    log_dir = os.path.join('logs', module_name)
    os.makedirs(log_dir, exist_ok=True)
    
    log_file = os.path.join(log_dir, f'{timestamp}.log')
    file_handler = logging.FileHandler(log_file)
    file_handler.setLevel(logging.INFO)
    
    formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
    file_handler.setFormatter(formatter)
    
    logger = logging.getLogger(module_name)
    logger.addHandler(file_handler)
    
    return logger

class AIClient:
    def __init__(self, module_name):
        self.client = OpenAI(
            base_url="https://openrouter.ai/api/v1",
            api_key="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"
        )
        self.logger = setup_logging(module_name)
        self.extra_headers = {
            "HTTP-Referer": "http://localhost",
            "X-Title": "NovelAI-Generator",
        }
        self.conversation_history = {}  # Store conversation history by conversation_id
        self.module_name = module_name
        
        # Setup conversation logging
        self.conversation_log_dir = os.path.join('logs', 'conversations')
        os.makedirs(self.conversation_log_dir, exist_ok=True)
        self.master_log_path = os.path.join('logs', 'master_conversation_log.json')
        self.master_log = self._load_or_create_master_log()

    def _load_or_create_master_log(self):
        """Load existing master log or create new one"""
        if os.path.exists(self.master_log_path):
            with open(self.master_log_path, 'r') as f:
                return json.load(f)
        return {
            "project_start_time": datetime.now().strftime('%Y%m%d_%H%M%S'),
            "conversations": {}
        }

    def _save_master_log(self):
        """Save master log to disk"""
        with open(self.master_log_path, 'w') as f:
            json.dump(self.master_log, f, indent=4)

    def _log_conversation(self, conversation_id, messages, response):
        """Log conversation to both individual file and master log"""
        timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
        
        # Create conversation entry
        conversation_entry = {
            "module": self.module_name,
            "timestamp": timestamp,
            "messages": messages,
            "response": response
        }
        
        # Save to individual conversation log
        conv_log_path = os.path.join(
            self.conversation_log_dir, 
            f'conversation_{conversation_id}_{timestamp}.json'
        )
        with open(conv_log_path, 'w') as f:
            json.dump(conversation_entry, f, indent=4)
        
        # Update master log
        if conversation_id not in self.master_log["conversations"]:
            self.master_log["conversations"][conversation_id] = []
        self.master_log["conversations"][conversation_id].append(conversation_entry)
        self._save_master_log()

    async def get_completion(self, messages, conversation_id=None, temperature=0.7):
        """Get completion from Claude API with conversation history"""
        try:
            # If conversation_id provided, append to existing history
            if conversation_id:
                if conversation_id not in self.conversation_history:
                    self.conversation_history[conversation_id] = []
                conversation_messages = self.conversation_history[conversation_id] + messages
            else:
                conversation_messages = messages

            completion = self.client.chat.completions.create(
                extra_headers=self.extra_headers,
                # model="anthropic/claude-3.5-sonnet:beta",
                model="deepseek/deepseek-r1",
                messages=conversation_messages,
                temperature=temperature
            )
            
            response = completion.choices[0].message.content
            
            # Update conversation history if conversation_id provided
            if conversation_id:
                self.conversation_history[conversation_id].extend([
                    messages[-1],  # Add user's message
                    {"role": "assistant", "content": response}  # Add AI's response
                ])
                # Log conversation update
                self.logger.info(f"Updated conversation {conversation_id} history. Current length: {len(self.conversation_history[conversation_id])}")
            
            # Log the interaction
            self._log_conversation(
                conversation_id or "single_interaction",
                messages,
                response
            )
            
            self.logger.info(f"Generated response for prompt: {messages[-1]['content'][:100]}...")
            return response
            
        except Exception as e:
            self.logger.error(f"Error getting completion: {str(e)}")
            raise

    def start_conversation(self):
        """Start a new conversation and return its ID"""
        conversation_id = datetime.now().strftime('%Y%m%d_%H%M%S')
        self.conversation_history[conversation_id] = []
        self.logger.info(f"Started new conversation: {conversation_id}")
        return conversation_id

    def end_conversation(self, conversation_id):
        """End a conversation and clean up its history"""
        if conversation_id in self.conversation_history:
            # Log final conversation state before ending
            self._log_conversation(
                conversation_id,
                self.conversation_history[conversation_id],
                "CONVERSATION_ENDED"
            )
            del self.conversation_history[conversation_id]
            self.logger.info(f"Ended conversation: {conversation_id}")

    def get_conversation_history(self, conversation_id):
        """Get the history of a specific conversation"""
        return self.conversation_history.get(conversation_id, [])

    def clear_conversation_history(self, conversation_id=None):
        """Clear conversation history for one or all conversations"""
        if conversation_id:
            if conversation_id in self.conversation_history:
                self.conversation_history[conversation_id] = []
                self.logger.info(f"Cleared history for conversation: {conversation_id}")
        else:
            self.conversation_history = {}
            self.logger.info("Cleared all conversation histories")
